bp_common/src/include/bp_common_me_if.vh

```diff
@@ -19,52 +19,46 @@
- * 5a. In the case of reading from memory, the CCE will retrieve the block then send both a
- *     bp_cce_lce_cmd_s to set the tag and bp_lce_data_cmd_s to provide the data block. Once
- *     the LCE receives both messages, the LCE sends a coherence ack in a bp_lce_cce_resp_s message
- *     to the CCE. The cache miss is complete and the LCE is now "woken up".
- * 5b. In the case of an LCE to LCE transfer, the CCE sends a bp_cce_lce_cmd_s to the LCE that
- *     will provide the block to initiate a transfer to the requesting LCE, and then sends a
- *     bp_cce_lce_cmd_s set tag message to the requesting LCE. The LCE sending the transfer sends
- *     a bp_lce_data_cmd_s to supply the requested data. Once the cache-missing LCE receives
- *     both the transfer response and set tag command, it sends a bp_lce_cce_resp_s to ack the
- *     receipt of the data transfer and set tag command, and it considers itself "woken up".
- * 6. If an LCE to LCE transfer occurred, the CCE will also perform a writeback of the block
- *    that was in the source LCE of the transfer. It does this by sending a bp_cce_lce_cmd_s
- *    for writeback and then waits for a bp_lce_cce_data_resp_s. The CCE finishes the writeback to
- *    the memory.
- *
- * LCE Input Message Priorities (highest to lowest)
- * 1. bp_lce_data_cmd_s
- * 2. bp_cce_lce_cmd_s
- *
- * CCE Input Message Priorities
- * 1. bp_mem_cce_data_resp_s
- * 2. bp_mem_cce_resp_s
- * 3. bp_lce_cce_req_s
- *
- *
- * Clients should use declare_bp_me_if(addr_width_mp, data_width_mp, num_lce_mp, lce_assoc_mp)
- *   to declare all me_if structs at once.
+ * There are 5 logical networks/message types:
+ * 1. LCE Request
+ * 2. LCE Response
+ * 3. LCE Data Response
+ * 4. LCE Command
+ * 5. LCE Data Command
+ *
+ * These five logical message types are carried on three networks:
+ * 1. Request (low priority)
+ * 2. Command (medium priority), LCE Commands and Data Commands
+ * 3. Response (high priority), LCE Responses and Data Responses
+ *
+ * A Request message may cause a Command message, and a Command message may cause a Response.
+ * A higher priority message may not cause a lower priority message to be sent, which avoids
+ * a circular dependency between message classes.
+ *
+ * LCE Request Processing Flow:
+ *  At a high level, a cache miss is handled by an LCE Request being sent to the CCE, followed by
+ *  a series of commands and and responses that handle invalidating, evicting, and writing-back
+ *  blocks as needed, sending data and tags to the LCE, and concluding with the LCE sending a response
+ *  to the CCE managing the transaction. The length of a coherence transaction depends on the type of
+ *  request (read- or write-miss), the current state of the requested block, and the current state of
+ *  the cache way that the miss will be filled into.
+ *
+ *
+ * Clients should use the declare_bp_me_if() macro to declare all of the interface structs at once.
+ *
  */
 
+
+
+/* TODO list
+
```

Comment:
For the first point, do you mean requiring cache block aligned transfers? I think that's a reasonable requirement

Why would cache block size being a multiple of flits help in the LCE-CCE-MEM case? I would think that the 'extra' flit data would just be zeros

Reply:
I mean having the first bit of the data be the first bit of some flit. The size of flits doesn't matter a whole lot actually, but I like powers of two :)

What I think we don't want is having the first bit of data be, e.g., bit 5 in a flit. The headers for data messages between LCE-CCE and CCE-MEM are likely to be different, so we don't really want any flit to contain both header and data info. We want some header flit(s), then data flits, where we can easily strip the header flits, create a new outbound header, then pass the data flits along effectively without touching them.

Reply:
Oh yep, definitely agree!  For ease of implementation, maybe standardize 1 header flit padded to data flit size (e.g. width of link). Do you think we need any headers > 32 bits?

Reply:
On the Command network we should be less than 32 bits. On the Response network, there is an address field, which is 39-bits by default so we probably need two 32-bit flits for the header. It might be possible to get rid of the address field though. For some messages the CCE might not care, and for others it might know what address the response is for based on its (CCE's) state.

Reply:
What could the address be used for? Would it make sense to encode that behavior as metadata?